### ISSUE TITLE
Community Call, September 2020, Agenda

### ISSUE THREAD

**Issue by larshp** [abapGit/abapGit#3834]
Time: 23rd Sept, 15:00 CEST

Duration: 1 hour scheduled

Invitations: Open for all, message me your email address on [Slack](https://abapgit-slackinviter.herokuapp.com/), and I'll forward you the invite

No recording, but minutes of the meeting will be available afterwards

**Agenda**

1. Welcome
2. Deprecating 2FA, #3150
3. Moving abapGit to the abapGit organization
4. Harmonization of object serializers in ABAP in the Cloud and abapGit. #3349
5. Local setup to meet abapGit's abaplint criterion -> [Separate call](https://github.com/larshp/abapGit/issues/3931) scheduled
6. PR discussion vs Issue discussion, [contribution guidelines](https://github.com/larshp/abapGit/blob/master/CONTRIBUTING.md)
7. table data serialization concept, #3441

**Comment by filak-sap** [comment]
If I don't make it to attend the call, don't forget that the `add last change date to object api, don't reserialize unchanged objects` is not OK in Decentralized development - I think it's not a good idea at all as time stamps are extremely fragile - we should rather convince ABAP creators to add object HASH to TADIR (impossible but the only reliable solution).

**Comment by larshp** [comment]
yea, some objects does not have timestamps like dynpros, ie. these must always be serialialized, so something might have a timestamp something might not, some might not have it implemented => a lot of complexity

Plus how to force serialization if the serializer code changes, then there must be another option on top of this

rather spend the time on trying to optimize serialization performance generally, and remember this already runs in parallel, so its possible to throw more hardware at it

I also think it would be a lot easier, assuming the objects are transported, to only serialize objects in open transports or released within the last week.

**Comment by mkaesemann** [comment]
@larshp: I started evaluating only serializing objects in open transports, but unfortunately that only works if the serializers are modified from the format they are using today, since some things (e.g. changes in includes) also cause changes in other objects (e.g. other structures or tables that include the include). Technically those dependent objects would not need to be serialized/commited, because deserialization and activation would trigger corresponding required changes changes when the code is pulled in another system, but the change detection and listing becomes confusing.
So this only works if the serializers for all objects produce an output format that is self-contained for all objects (e.g. the representation of tables with includes that ADT gives in its table editor is self contained, while generating a complete field list with resolved includes and translating that to XML is not).

**Comment by larshp** [comment]
@mkaesemann yea, suggest opening an issue with an example each time this occurs

**Comment by mkaesemann** [comment]
That's the plan (along with other improvements I want to address). I'mtrying to get an internal project in our company going to implement these kinds of improvements to abapGit, but our daily business is crazy at the moment and it will probably have to be postponed until next year... :(

**Comment by pokrakam** [comment]
Re. timestamps, I think it's feasible, but must definitely be configurable (by repo?). 
e.g. it would help the person working on their own NPL system where everything is under their control, but not in a decentralised scenario. Some additional rules are needed, e.g. only timestamps prior today are considered. Persisting local hashes could also be an idea.

**Comment by sbcgua** [comment]
Assuming "table data serialization concept" is winning in the slack voting: what about serializing data as tab-delimited text ? https://github.com/sbcgua/text2tab. Imho XML would be too heavy ...

**Comment by flaiker** [comment]
Or just use cl_abap_container_utilities? Will have lots of whitespace but no alignment guesswork.

**Comment by FreHu** [comment]
I am a fan of json for serializing in general, I would even welcome json as a switchable alternative to all the current xml files.

As for the last change date, I guess this is overall not a good idea. At least the reasons are now documented :). I will rephrase my point as "how to improve serialization performance given that most objects are likely unchanged between refreshes"

**Comment by sbcgua** [comment]
"json as an alternative to all the current xml files" :+1: But we need to measure performance. I intuitively expect that XML (as  kernel supported format) is rather fast. kernel json serializer ? didn't measure it.

"last change date": heh, we had caching before, it gave good speed up, but had some issues from time to time and complicated architecture ... though I would say there should be a layer like that. Maybe switchable on/off.

**Comment by sbcgua** [comment]
csv (tabbed) vs json (for table dumps):
- json is more "diffable", but takes more space (attrs are repeated)
- csv - less diffable (though new tools can highlight in-line changes), theoretically more readable (to certain extent ... big tables would shift values related to headers) and openable in Excel (this is big, imho)

So quite depends on what is the goal of such serialization. Customizing ? maybe json is better. Data ? Probably csv

**Comment by mbtools** [comment]
i have "json instead of xml" also on my list. seems doable but has side effects for example on abaplint (reads the xmls) that need to be considered.

we will have to specify somewhere which tables and which data in the tables should be serialized. we might as well specify the format and allow for several adaptors.

**Comment by larshp** [comment]
[![image](https://user-images.githubusercontent.com/5888506/92447505-95a2b900-f1b7-11ea-85cf-5b9f1a1dd6a0.png)](https://www.timeanddate.com/worldclock/meetingdetails.html?year=2020&month=9&day=23&hour=13&min=0&sec=0&p1=69&p2=250&p3=240)

**Comment by larshp** [comment]
I've sent out invitations, let me know if you have not received it

The meeting is open for all, message me your email address on Slack, and I'll forward you the invite

**Comment by jrodriguez-rc** [comment]
Supporting serialization in json needs some requisitions for 702 and 731

> WIth Releases 7.02 and 7.03/7.31 (Kernelpatch 116) JSON is supported natively in ABAP
https://blogs.sap.com/2013/01/07/abap-and-json/

> Yes we had but since the JSON support was made available in Release 7.40 and downported to 7.02 and 7.31 (SAP Notes 1648418 and 1650141) it shouldn’t be missed in a systematic overview over all important ABAP News for Release 7.40
https://blogs.sap.com/2013/07/04/abap-news-for-release-740-abap-and-json/

**Comment by larshp** [comment]
next meeting -> https://github.com/abapGit/abapGit/issues/3946

closing
